File: ts/src/founder.ts
export class FounderTree {
  private founders: Map<string, string>; // dir_path → participant_id
  private fileFounders: Map<string, string>; // file_path → participant_id (file creator)
  private rootFounder: string;

  constructor(rootFounder: string) {
    this.founders = new Map();
    this.fileFounders = new Map();
    this.founders.set('.', rootFounder);
    this.rootFounder = rootFounder;
  }

  registerFromWrite(filePath: string, participant: string): void {
    // Extract directory from file path
    const lastSlash = filePath.lastIndexOf('/');
    if (lastSlash === -1) {
      return; // No directory (root level)
    }
    const dir = filePath.substring(0, lastSlash);

    // Walk through parent directories and register if not exists
    let current = '';
    for (const segment of dir.split('/')) {
      if (segment === '') continue;

      if (current !== '') {
        current += '/';
      }
      current += segment;

      if (!this.founders.has(current)) {
        this.founders.set(current, participant);
      }
    }
  }

  getFounder(dir: string): string | null {
    return this.founders.get(dir) || null;
  }

  isAncestorFounder(upperDir: string, lowerDir: string): boolean {
    if (upperDir === '.') {
      return lowerDir !== '.';
    }
    return lowerDir.startsWith(upperDir) && lowerDir !== upperDir;
  }

  listAll(): Array<[string, string]> {
    const entries = Array.from(this.founders.entries());
    entries.sort((a, b) => a[0].localeCompare(b[0]));
    return entries;
  }

  getRootFounder(): string {
    return this.rootFounder;
  }

  getDirFounderForFile(filePath: string): string | null {
    const lastSlash = filePath.lastIndexOf('/');
    const dir = lastSlash === -1 ? '.' : filePath.substring(0, lastSlash);
    return this.getFounder(dir);
  }

  registerFileFounder(filePath: string, participant: string): void {
    if (!this.fileFounders.has(filePath)) {
      this.fileFounders.set(filePath, participant);
    }
  }

  getFileFounder(filePath: string): string | null {
    return this.fileFounders.get(filePath) || null;
  }

  canAcceptDelete(filePath: string, participant: string): boolean {
    const fileFounder = this.getFileFounder(filePath);
    return fileFounder === participant;
  }

  toJSON(): string {
    const obj: Record<string, string> = {};
    for (const [k, v] of this.founders.entries()) {
      obj[k] = v;
    }
    return JSON.stringify(obj);
  }

  getFoundersObject(): Record<string, string> {
    const obj: Record<string, string> = {};
    for (const [k, v] of this.founders.entries()) {
      obj[k] = v;
    }
    return obj;
  }

  loadFoundersObject(obj: Record<string, string>, fileFoundersObj?: Record<string, string>) {
    this.founders = new Map(Object.entries(obj));
    if (fileFoundersObj) {
      this.fileFounders = new Map(Object.entries(fileFoundersObj));
    }
  }

  static fromJSON(json: string, rootFounder: string): FounderTree {
    const obj = JSON.parse(json);
    const tree = new FounderTree(rootFounder);
    tree.founders = new Map(Object.entries(obj));
    return tree;
  }
}

File: rust/src/founder.rs
use std::collections::HashMap;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub struct FounderTree {
    founders: HashMap<String, String>, // dir_path → participant_id
    root_founder: String,
}

impl FounderTree {
    pub fn new(root_founder: &str) -> Self {
        let mut founders = HashMap::new();
        founders.insert(".".to_string(), root_founder.to_string());
        Self {
            founders,
            root_founder: root_founder.to_string(),
        }
    }

    /// write 時にファイルのディレクトリ階層を走査し Founder を自動登録
    pub fn register_from_write(&mut self, file_path: &str, participant: &str) {
        // ファイルパスからディレクトリを抽出
        let dir = match file_path.rfind('/') {
            Some(pos) => &file_path[..pos],
            None => return, // ディレクトリなし（root直下）
        };

        // 各親ディレクトリを走査して未登録なら participant を登録
        let mut current = String::new();
        for segment in dir.split('/') {
            if segment.is_empty() {
                continue;
            }
            if !current.is_empty() {
                current.push('/');
            }
            current.push_str(segment);

            self.founders
                .entry(current.clone())
                .or_insert_with(|| participant.to_string());
        }
    }

    pub fn get_founder(&self, dir: &str) -> Option<&str> {
        self.founders.get(dir).map(|s| s.as_str())
    }

    /// 上位 Founder が下位 Founder の Leader かどうか
    /// upper_dir が lower_dir の親ディレクトリであればtrue
    pub fn is_ancestor_founder(&self, upper_dir: &str, lower_dir: &str) -> bool {
        if upper_dir == "." {
            // root は全ディレクトリの親
            return lower_dir != ".";
        }
        lower_dir.starts_with(upper_dir) && lower_dir != upper_dir
    }

    pub fn list_all(&self) -> Vec<(&str, &str)> {
        let mut entries: Vec<_> = self
            .founders
            .iter()
            .map(|(k, v)| (k.as_str(), v.as_str()))
            .collect();
        entries.sort_by_key(|(k, _)| k.to_string());
        entries
    }

    pub fn get_root_founder(&self) -> &str {
        &self.root_founder
    }

    /// ファイルが属するディレクトリの Founder を取得
    pub fn get_file_founder(&self, file_path: &str) -> Option<&str> {
        let dir = match file_path.rfind('/') {
            Some(pos) => &file_path[..pos],
            None => ".", // root直下
        };
        self.get_founder(dir)
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string(&self.founders).unwrap()
    }

    pub fn from_json(json: &str, root_founder: &str) -> Self {
        let founders: HashMap<String, String> = serde_json::from_str(json).unwrap_or_else(|_| HashMap::new());
        Self {
            founders,
            root_founder: root_founder.to_string(),
        }
    }
}
